infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java

```diff
@@ -0,0 +1,197 @@
+
+        if (this.handshakeFailed.get()) {
+            return;
+        }
+
+        if (this.handshakeComplete.get()) {
+            // If handshake completed successfully, but still a message came through this handler,
+            // send on to the next handler in order to avoid message loss.
+            super.channelRead(ctx, m);
+            return;
+        }
+
+        CorfuPayloadMsg<HandshakeMsg> handshake;
+
+        try {
+            handshake = (CorfuPayloadMsg<HandshakeMsg>) m;
+            log.debug("channelRead: Handshake Message received. Removing readTimeoutHandler " +
+                    "from pipeline.");
+            ctx.pipeline().remove("readTimeoutHandler");
+        } catch (ClassCastException e) {
+            log.warn("channelRead: Non-handshake message received by handshake handler." +
+                    " Send upstream only if handshake succeeded.");
+            if (this.handshakeComplete.get()) {
+                // Only send upstream if handshake is complete.
+                super.channelRead(ctx, m);
+            }
+            return;
+        }
+
+        UUID clientId = handshake.getPayload().getClientId();
+        UUID serverId = handshake.getPayload().getServerId();
+
+        // Validate handshake
+        if (!serverId.equals(this.nodeId)) {
+            log.error("channelRead: Invalid handshake: this is " + this.nodeId +
+                    " and client is trying to connect to " + serverId);
+            this.fireHandshakeFailed(ctx);
+            return;
+        }
+
+        // Store clientID as a channel attribute.
+        ctx.channel().attr(clientIdAttrKey).set(clientId);
+        log.info("channelRead: Handshake validated by Server.");
+        log.debug("channelRead: Sending handshake response: Node Id: " + this.nodeId +
+                " Corfu Version: " + this.corfuVersion);
+
+        CorfuMsg handshakeResponse = CorfuMsgType.HANDSHAKE_RESPONSE
+                .payloadMsg(new HandshakeResponse(this.nodeId, this.corfuVersion));
+        ctx.writeAndFlush(handshakeResponse);
+
+        // Flush messages in queue
+        log.debug("channelRead: There are {" + this.messages.size() + "} messages in queue to " +
+                "be flushed.");
+        for (CorfuMsg message : this.messages) {
+            ctx.writeAndFlush(message);
+        }
+
+        // Remove this handler from the pipeline; handshake is completed.
+        log.info("channelRead: Removing handshake handler from pipeline.");
+        ctx.pipeline().remove(this);
+        this.fireHandshakeSucceeded();
+    }
+
+    @Override
+    public void channelActive(final ChannelHandlerContext ctx) throws Exception {
+        log.info("channelActive: Incoming connection established from: " +
+                ctx.channel().remoteAddress() + " Start Read Timeout.");
+        ctx.pipeline().addBefore(ctx.name(), "readTimeoutHandler",
+                new ReadTimeoutHandler(10));
```

Comment:
should 10 be a constant or a config property ?

Reply:
Hm, yeah it should be a config property. Can you pull this from the options map in ServerContext?

Reply:
Done